runtime/doc/lsp.txt

```diff
@@ -698,6 +715,75 @@ commands                                                    *vim.lsp.commands*
 
     The second argument is the `ctx` of |lsp-handler|
 
+config({name}, {cfg})                                       *vim.lsp.config()*
+    Define default configuration for LSP clients.
+
+    Use name '*' to set default configuration for all clients.
+
+    Can also be used as a table.
+
+    Examples:
+    • Set default root marker for all clients: >lua
+       vim.lsp.config('*', {
+         root_markers = { '.git' },
+       })
+<
+    • Set default configuration for clangd using table form: >lua
+       vim.lsp.config.clangd = {
+         cmd = {
+           'clangd',
+           '--clang-tidy',
+           '--background-index',
+           '--offset-encoding=utf-8',
+         },
+         markers = { '.clangd', 'compile_commands.json' },
+         filetypes = { 'c', 'cpp' },
+       }
+<
+    • Get configuration for luals: >lua
+       local cfg = vim.lsp.config.luals
+<
+
+    Parameters: ~
+      • {name}  (`string`)
+      • {cfg}   (`vim.lsp.Config`) See |vim.lsp.Config|.
+
+enable({name}, {config})                                    *vim.lsp.enable()*
```

Comment:
If there is potential that people will want to "disable" configs, then `{config}` arg will be able to take `false` in the future?

Reply:
Do we even need a `config` arg on `enable()` or could we require a `vim.lsp.config()` call?

```lua
vim.lsp.config('foo', {...})
vim.lsp.enable('foo')
```

Reply:
It's up to you. I'm happy to allow a config arg to avoid boilerplate which is one of the goals of this PR.

We can allow `false` to be passed to disable a config. Do you want me to add that in this PR or can it be deferred?

Reply:
> We can allow `false` to be passed to disable a config. Do you want me to add that in this PR or can it be deferred?


Can be deferred, just wanted to make sure the door is open for it. Might be worth a 1-line TODO comment.



> I'm happy to allow a config arg to avoid boilerplate which is one of the goals of this PR.

Most users that are using the default configs will only need to call `enable()` (no config-arg), right? I think it's perfectly find to require a `vim.lsp.config()` call for overriding defaults or redefining a config. And that cleanly separates the purpose of each, and also simplifies the docs.

Reply:
What's the plan for distributing default configs to users? Is it the rtp approach? Otherwise, when this is merged, and potentially for the 0.11 release, the common usecase will be providing `config()` and `enable()` together.

Reply:
> the common usecase will be providing `config()` and `enable()` together.

Wouldn't that imply we are not providing any default configs? We do plan to provide default configs somehow. Perhaps starting by modifying nvim-lspconfig, and then upstreaming popular configs.

Reply:
> then Nvim itself does the config() call.

When/where? Via the rtp method via a future PR?

Reply:
One usecase that I've tried to cover is something like:

```lua
vim.lsp.enable('bashls', {
  filetypes = 'rando_bash_ft'
  settings = {
    bashIde = {
      shellcheckArguments = {
        '-e', 'SC2086', -- Double quote to prevent globbing and word splitting
        '-e', 'SC2155', -- Declare and assign separately to avoid masking return values
      },
    }
  }
})
```

Where the passed config is partial and allows easily overriding default values.

If we require `vim.lsp.config()` for this then it becomes:

```lua
vim.lsp.config.bashls = vim.tbl_deep_extend('force', vim.lsp.config.bashls, {
  settings = {
    bashIde = {
      shellcheckArguments = {
        '-e', 'SC2086', -- Double quote to prevent globbing and word splitting
        '-e', 'SC2155', -- Declare and assign separately to avoid masking return values
      },
    }
  }
})

table.insert(vim.lsp.config.bashls.filetypes, 'rando_bash_ft')

vim.lsp.enable('bashls')
```

Not a deal breaker, but worth noting.

Reply:
> However we decide to do it, it will involve a config() call, right?

Yes, but also changes to `vim.lsp.enable()`.

I'm mainly asking because I think forcing `vim.lsp.config()` accelerates the need for us (or plugins) distributing default configs **efficiently**. If we want to defer that, then allowing config in `vim.lsp.enable` is a nice reduction in friction for the meantime.

Reply:
Shipping default configs keeps coming up, but do we have some concrete examples of what that would entail?

For example, looking at `clangd` in lspconfig:

- cmd: `clangd`
- filetypes: `filetypes = { 'c', 'cpp', 'objc', 'objcpp', 'cuda', 'proto' },`
- root_dir: ` root_dir = function(fname)
      return util.root_pattern(
        '.clangd',
        '.clang-tidy',
        '.clang-format',
        'compile_commands.json',
        'compile_flags.txt',
        'configure.ac' -- AutoTools
      )(fname) or util.find_git_ancestor(fname)
    end,`


- `single_file_support`: Not in core, imho shouldn't be, that's `root_dir = nil`
- capabilities: These are off-spec definitions, not in scope of score
- commands: user commands, shouldn't be part of lsp interfaces. The functions they enable are also off-spec


---

I'm still unsure if defining root dir heuristics for the users is a good move as it can lead to surprising behavior. In vscode and all the other "easy to use" editors the choice is also made by the user.
In helix there are also issues about it doing the wrong thing.

_If_ we include them, we'd imho have to think how to surface it more prominently (e.g. different statusline defaults)

---

Speaking for how I'd use it, I'd use both lsp.config and lsp.enable pretty much 100% of the time.

Reply:
A lot of users will want to pull in these settings from somewhere even if it's just for command and filetypes.

I'm less concerned about including these in core, but more about enabling a general mechanism so plugins can provide these configurations passively. At least then this logic can be deferred to someone else to figure out instead of the user.

And obviously it's worth reiterating that the goal isn't a one size fits all, just most.

I also feel like you potentially attribute a lot more weight to potential issues here because you use java? For all the languages I use, this model fits very well. And for cases it doesn't, there's appropriate escape hatches to fix things.

Reply:
> Oh that's definitely less ergonomic. I guess you already considered this, but the obvious question is why can't we make config() have the same ergonomics as your enable() example?

Like your example, we can make `vim.lsp.config()` update a config, whereas `vim.lsp.config[name] = {}` completely redefines a config.

In the PR's current form, the config passed to `enable` affects the global config, but is lazy resolved. Not a big deal. 

Also something I've been trying to keep in mind, but haven't been able to implement is accommodating project specific instances of `vim.lsp.enable()`. But I haven't got anything, so lets go ahead with the simplification. I'll also remove the `*` configuration since that's not that useful in practice.

Reply:
> I'll also remove the * configuration since that's not that useful in practice.

It would allow plugins that add extra capabilities not yet in core to set `lsp.config("*", { capabilities = ... })`

> I also feel like you potentially attribute a lot more weight to potential issues here because you use java? For all the languages I use, this model fits very well. And for cases it doesn't, there's appropriate escape hatches to fix things.

java is certainly a special snowflake but the issue with root dir detection is not limited to it.
There is a good chunk of closed issues in lspconfig related to servers not attaching, or starting more than one, or slow performance - all related in some way to root dir definitions. 
Murphy's law: Anything that can go wrong will go wrong. 
If users set it explicitly themselves they'd at least have some lead of what could've led to the issue. I agree that the goal isn't one size fits all, but if we do make a choice for users we need to think about how it affects users where the choice doesn't fit and how they can deal with it. One option I can think of is making root_dir choice and server count visible in statusline and maybe clickable to lead them to the health check or the docs for more info.
(Or simply not make the choice for them)

I raise this only as a concern for moving the root markers into core.

What I agree with or see as unproblematic:
- configs should be mergeable in some form, and plugins should be able to set them passivley. (Especially capabilities, since they have to be set before a client/server starts)
- server name -> filetypes mapping could be in core, that's less likely to cause issues. (In vscode that's something the language extensions define) although the "make it visible" aspect still applies somewhat.


> Is there really much issue with having vim.lsp.config() separated out (which does config merging)? It does simplify how to explain this.

No big deal, but I already see myself adding some kind of `enable(name, config)` that would call `lsp.config(name, config); lsp.enable(name)`  to avoid the `lsp.config(); lsp.enable() * server` pairs - so why not provide it?

Reply:
> Shipping default configs keeps coming up

Yes, and partly my fault, so let me state clearly here what I meant by that:
1. We **want** to ship default configs with `nvim-lspconfig` (as that is that repo's purpose; i.e., the goal is not shipping something more but _removing everything else_).
2. We _then_ **might** want to ship default configs with Neovim for languages we think should be easy to set up out-of-the-box _but currently aren't_ -- that list may very well be empty (but may not necessarily stay that way). Potential candidates are `clangd` (too simple), `luals` (already handled by our editorconfig), `ts_query_ls` (not done yet).

TL;DR: the main target for "shipping configs" is nvim-lspconfig, _not_ core; that is just a door we're leaving open (at no current cost).

Reply:
> No big deal, but I already see myself adding some kind of enable(name, config) that would call lsp.config(name, config); lsp.enable(name) to avoid the lsp.config(); lsp.enable() * server pairs - so why not provide it?

Me too, and hence my initial pushback to Justin's comments. I agree with Justin that It does simplify the model and makes the accountabilities of each function very clear. And it's not a closed door to allow this in the future.

Reply:
And given that the "real" target here is nvim-lsponfig, that strongly pushes us towards passive discovery on `rtp` because otherwise we'd need to keep some sort of setup logic in `nvim-lspconfig`, again duplicating logic. `rtp` precedence is also well-documented and (despite some hitches) established pattern for custom and overridden treesitter queries.

(If we go that route, an open question is about _merging_ matching configs on `rtp`, but I'm not convinced that needs to or should be part of the initial implementation. Worth keeping in mind, though.)

Reply:
> (If we go that route, an open question is about merging matching configs on rtp, but I'm not convinced that needs to or should be part of the initial implementation. Worth keeping in mind, though.)

We wouldn't have to merge. We can just source the files according to rtp rules. Each file can either call `vim.lsp.config()` to update a config, or use `vim.lsp.config[name] = ...` to redefine a config. It'll be down to the file what they want to do.

And to note the merging semantics I've added here will be `tbl_deep_extend` except:
- Lists: `filetypes` and `root_markers` are extended.
- Fields can be single or list (`on_attach`, `on_exit`, `on_init`), so they are coerced into a list.

Reply:
Even better.

Reply:
> > Is there really much issue with having vim.lsp.config() separated out (which does config merging)? It does simplify how to explain this.
> 
> No big deal, but I already see myself adding some kind of `enable(name, config)` that would call `lsp.config(name, config); lsp.enable(name)` to avoid the `lsp.config(); lsp.enable() * server` pairs - so why not provide it?

Because
1. it's confusing to explain (and understand, for users) how `config('foo', <config>)` differs from `enable('foo', <config>)`. If `config()` owns setting/merging configs then docs and concepts are simpler. That's gain is worth the cost of one extra function call (which will not even be needed for users that don't want to modify the default 'foo' config).
2. the door is open to extending `enable()` in a future PR.